### PR TITLE
Remove filter to show only gems that are not yanked in search.

### LINKED ISSUE #2110: Mention yanked versions exist in search
I've recently released a new gem called Dawn.

As I did in the past, I checked on RubyGems before I even picked the name, to make sure it is available on it's purest form (i.e. just "Dawn"). Since no result was found, I happily went on with it and released it. I then checked out my new RubyGems page, and found out I have version history, which I clearly do not... It turned out there was a gem called "Dawn" before, and it has been yanked, thus it did not appear on the search results.

Had I known the name was used in the past by a gem which is now yanked, I might have chosen a different name, to avoid confusion.

Perhaps search should not hide yanked gems completely, but mention they do exist in some manner?

### PR DESCRIPTION
**Pull request by vachhanihpavan**
Fixes #2110 
Remove the filter which was blocking the yanked gems to appear in the search results.
Please give suggestions if anything more is required in UI.

### CODE REVIEW

**Comment by simi** [comment]
@vachhanihpavan can you provide screenshot? For example, is it clear gem is completely yanked in search result?

**Comment by vachhanihpavan** [comment]
Hi @simi, I haven't done any changes in views to show if the gem is yanked. So it appears same as normal gem result. 
Screenshot from 2020-03-13 15-49-55
For eg. here I yanked activesupport gem in my local setup. But it is still there in the search results.

Please give suggestions on how to make UI more intuitive.
Thanks :smile:

**Comment by vachhanihpavan** [comment]
Screenshot from 2020-03-13 18-03-47

Hey @simi, I have added a badge to show if the gem is yanked.

**Comment by colby-swandale** [comment]
Thanks for working on this PR but I am not sold on this idea from #2110

To me, showing yanked gems in the search results would be counter intuitive for regular users as the gem is in a state where the user cannot view, download or install the gem. 

Here's a few gems that i yanked a few weeks ago:

 * https://rubygems.org/gems/ebay-client
 * https://rubygems.org/gems/logstash_codec-avro
 * https://rubygems.org/gems/cwlogs_s3

An alternative solution i think would be add a recommendation into the guides for gem authors to first check if the gem already exists in rubygems.org, ie: going to `http://rubygems.org/gems/<name>`. And updating the page to explicitly say something along "This gem and all versions have been yanked" instead of "This gem is not currently hosted on RubyGems.org."

**Comment by sonalkr132** [comment]
>  would be counter intuitive for regular users as the gem is in a state where the user cannot view, download or install the gem

agreed :+1: lets not add it in regular search results. @colby-swandale would you say it is okay to show yanked gem after user has clicked on a `yanked` filter on the search page? We can keep the yanked label buts we won't show version, downloads or description. Description could be [not_hosted_notice](https://github.com/rubygems/rubygems.org/blob/2878cdbc73eb69f29b03d3902e029ee427c0a785/config/locales/en.yml#L325):
> This gem is not currently hosted on RubyGems.org

> add a recommendation into the guides for gem authors to first check if the gem already exists in rubygems.org

@vachhanihpavan are you interested in doing this?

**Comment by colby-swandale** [comment]
I like that 👍

**Comment by vachhanihpavan** [comment]
@sonalkr132 I moved somed helper methods to resolve the failing cop. https://github.com/vachhanihpavan/rubygems.org/pull/1
Can you please review if that is fine. This PR will be ready after.

File: app/assets/javascripts/search.js [hunk 1]
```diff
@@ -5,18 +5,21 @@ if($("#advanced-search").length){
...
     .add($downloads)
     .add($updated)
+    .add($yanked)
     .on('input', function(e) {
       var name        = $name.val().length > 0 ? 'name: ' + $name.val() : '';
       var summary     = $summary.val().length > 0 ? 'summary: ' + $summary.val() : '';
       var description = $description.val().length > 0 ? 'description: ' + $description.val() : '';
       var downloads   = $downloads.val().length > 0 ? 'downloads: ' + $downloads.val() : '';
       var updated     = $updated.val().length > 0 ? 'updated: ' + $updated.val() : '';
+      var yanked      = $yanked.is(":checked") ? 'yanked: true' : '';
```
Comment (sonalkr132) [inline app/assets/javascripts/search.js]: We would like to show count of yanked gems here. I think we need to add an aggregation like [this](https://github.com/rubygems/rubygems.org/blob/master/lib/elastic_searcher.rb#L47) with filter `yanked: true`.

**Comment by sonalkr132** [comment]
> vachhanihpavan#1

please open PR on rubygems.org master.
Also, I would probably move subscribe_link and unsubscribe_link to separate helper instead. nice_date_for and simple_markup could be moved to application helper.

**Comment by vachhanihpavan** [comment]
> Also, I would probably move subscribe_link and unsubscribe_link to separate helper instead. nice_date_for and simple_markup could be moved to application helper.

Should I make a `links_helper.rb` and shift `subscribe_link`, `unsubscribe_link`, `reverse_dependencies_link`, `badge_link` and `report_abuse_link` there?

File: app/assets/javascripts/search.js (hunk 1 above)
Comment (vachhanihpavan) [reply app/assets/javascripts/search.js]: Hey, so as discussed above, we will show yanked gems, if a filter is checked. So if user is not shown any yanked gems in normal result but is given a filter with count of yanked gem will make it improper imo. Because currrently, the counts of filters are out of the shown results only.
Also if the user has selected the yanked filter, all the shown gems will be yanked, show again filter doesn't make sense.

Can you clarify this? @sonalkr132 :sweat_smile:

Comment (sonalkr132) [reply app/assets/javascripts/search.js:21]: I don't think filters have to co-related with results on search page. Consider example of github search, it shows search result of code by default, however if you click on filter of `issues`, it will show results only for that filter.
Issues, commits etc are not part of default search result page but their count is helpful info.
Perhaps I can better understand your concern if you can further explain why you believe this could be an issue.

File: test/unit/rubygem_searchable_test.rb [hunk 2]
```diff
@@ -83,7 +83,7 @@ class RubygemSearchableTest < ActiveSupport::TestCase
       import_and_refresh
     end
 
-    should "filter yanked gems from the result" do
+    should "show yanked gems in the result" do
```
Comment (sonalkr132) [inline test/unit/rubygem_searchable_test.rb:86]: I don't think we need this change?

File: app/views/rubygems/_rubygem.html.erb [hunk 3]
```diff
@@ -1,15 +1,25 @@
 <%= link_to rubygem_path(rubygem.name), :class => 'gems__gem' do %>
-  <span class="gems__gem__info">
+  <% if yanked?(rubygem) %>
+    <span class="gems__gem__info">
+    <h2 class="gems__gem__name">
+      <%= rubygem.name %>
+      <span class="gems__gem__badge">Yanked</span>
```
Comment (sonalkr132) [inline app/views/rubygems/_rubygem.html.erb:6]: i18n

File: lib/elastic_searcher.rb [hunk 4]
```diff
@@ -59,6 +58,9 @@ def search_definition # rubocop:disable Metrics/MethodLength
         end
       end
 
+      # only return gems that are not yanked unless filter is used
+      post_filter { term yanked: false } unless query_str.include? "yanked:true"
```
Comment (sonalkr132) [inline lib/elastic_searcher.rb:62]: is there a way we don't have to hardcode `yanked:true`? filter can also be written as `yanked: true`. ruby-es may have option for parsed query.

File: app/views/rubygems/_rubygem.html.erb [hunk 5]
```diff
@@ -1,15 +1,25 @@
...
-  <p class="gems__gem__downloads__count">
-    <%= download_count rubygem %>
-    <span class="gems__gem__downloads__heading">
+    <p class="gems__gem__downloads__count">
+      <%= download_count rubygem %>
+      <span class="gems__gem__downloads__heading">
       <%= t('rubygems.index.downloads') %>
     </span>
-  </p>
+    </p>
```
Comment (sonalkr132) [inline app/views/rubygems/_rubygem.html.erb:23]: Please fix indentation of [these lines](https://github.com/rubygems/rubygems.org/blob/9720ac7932faa36b27bf22c5799955a82e2c52a9/app/views/rubygems/_rubygem.html.erb#L11-L23)

File: app/helpers/searches_helper.rb [hunk 6]
```diff
@@ -26,4 +26,12 @@ def aggregation_count(aggregration, duration, buckets_pos)
     update_info = (duration == 30.days ? t("searches.show.month_update", count: count) : t("searches.show.week_update", count: count))
     link_to update_info, path, class: "t-link--black"
   end
+
+  def aggregation_bool_count(aggregration, field, filter)
+    count = aggregration["buckets"][field]["doc_count"]
+    return unless count > 0 && !params[:query].include?("#{field}:#{filter}")
```
Comment (sonalkr132) [inline app/helpers/searches_helper.rb:32]: why we do need `!params[:query].include?("#{field}:#{filter}")`?

File: app/assets/javascripts/search.js [hunk 7]
```diff
@@ -5,18 +5,21 @@ if($("#advanced-search").length){
...
 
-      $main.val($.trim(name + ' ' + summary + ' ' + description + ' ' + downloads + ' ' + updated));
+        $main.val($.trim(name + ' ' + summary + ' ' + description + ' ' + downloads + ' ' + updated + ' ' + yanked));
```
Comment (sonalkr132) [inline app/assets/javascripts/search.js:23]: please don't change indentation, here and on line 10. make sure you editor is set for picking indentation automatically. we have used 2 spaces in all files.

File: app/helpers/searches_helper.rb (hunk 6 above)
Comment (vachhanihpavan) [reply app/helpers/searches_helper.rb:32]: Because if yanked filter is applied, we should not show `YANKED(123)`, because all the results are yanked only.

File: app/views/rubygems/_rubygem.html.erb (hunk 5 above)
Comment (vachhanihpavan) [reply app/views/rubygems/_rubygem.html.erb:23]: I added an if statement so, this intendation was added.

File: app/views/rubygems/_rubygem.html.erb (hunk 3 above)
Comment (vachhanihpavan) [reply app/views/rubygems/_rubygem.html.erb:6]: Can I reuse `searches.advanced.yanked` here also or should I create a new key under `rubygems` key?

File: app/views/rubygems/_rubygem.html.erb (hunk 5 above)
Comment (sonalkr132) [reply app/views/rubygems/_rubygem.html.erb:23]: check the link i have mentioned. to be specific, make sure span opening and closing tags are alligned, h2 is indented within span and bottom span content is indented.

File: app/helpers/searches_helper.rb (hunk 6 above)
Comment (sonalkr132) [reply app/helpers/searches_helper.rb:32]: >  we should not show YANKED(123)

we don't this for rest of the filters to keep it clear to the user that result shown is for the specific filter (not default result). imagine link being shared with the filter.

File: app/views/rubygems/_rubygem.html.erb (hunk 3 above)
Comment (sonalkr132) [reply app/views/rubygems/_rubygem.html.erb:6]: yes, reuse is fine.

File: app/helpers/searches_helper.rb (hunk 6 above)
Comment (vachhanihpavan) [reply app/helpers/searches_helper.rb:32]: If `yanked: true` filter is already applied by the user, without `!params[:query].include?("#{field}:#{filter}")` filter link will be visible and link will keep appending the same filter as
`search?query=name%3A+elasticsearch+++++yanked%3Atrue+AND+yanked%3Atrue`

Comment (sonalkr132) [reply app/helpers/searches_helper.rb:32]: >  link will keep appending the same filter

this shouldn't affect user experience in any away. If you really care, you can probably update the url to use absolute path , I think `url_for` will avoid this.
I hope you agree that we should keep it clear to the user, whether the page is for default results or one with a filter. Please let me know if there is any confusion, I would be happy to explain further :slightly_smiling_face:

File: app/assets/javascripts/search.js [hunk 8]
```diff
@@ -5,18 +5,21 @@ if($("#advanced-search").length){
...
-      var updated     = $updated.val().length > 0 ? 'updated: ' + $updated.val() : '';
+      var name        = $name.val().length > 0 ? 'name: ' + $name.val() + ' ' : '';
+      var summary     = $summary.val().length > 0 ? 'summary: ' + $summary.val() + ' ' : '';
+      var description = $description.val().length > 0 ? 'description: ' + $description.val() + ' ' : '';
+      var downloads   = $downloads.val().length > 0 ? 'downloads: ' + $downloads.val() + ' ' : '';
+      var updated     = $updated.val().length > 0 ? 'updated: ' + $updated.val() + ' ' : '';
+      var yanked      = $yanked.is(":checked") ? 'yanked: true' : '';
 
-      $main.val($.trim(name + ' ' + summary + ' ' + description + ' ' + downloads + ' ' + updated));
+      $main.val($.trim(name + summary + description + downloads + updated + yanked));
```
Comment (sonalkr132) [inline app/assets/javascripts/search.js:23]: why are we removing space again? if it is not related to yanked search result, can you please send a separate PR for it?

File: lib/elastic_searcher.rb [hunk 9]
```diff
@@ -59,7 +59,7 @@ def search_definition # rubocop:disable Metrics/MethodLength
       end
 
       # only return gems that are not yanked unless filter is used
-      post_filter { term yanked: false } unless query_str.include? "yanked:true"
+      post_filter { term yanked: false } unless query_str.delete(" ").include? "yanked:true"
```
Comment (sonalkr132) [inline lib/elastic_searcher.rb:62]: please use `gsub(/[[:space:]]/, "")`

Comment (vachhanihpavan) [reply lib/elastic_searcher.rb:62]: Isn't `delete(" ")` much faster for this purpose?
Any reason for `gsub`?

File: app/assets/javascripts/search.js (hunk 8 above)
Comment (vachhanihpavan) [reply app/assets/javascripts/search.js:23]: This just avoids extra spaces coming up in the query string.
For eg. if summary is not used n the filter, unnecessary space if added.
This doesn't affect the search results though. I just did this for JS cleanup that we discussed.

Comment (vachhanihpavan) [reply app/assets/javascripts/search.js:23]: Should I make a separate PR?

File: lib/elastic_searcher.rb (hunk 9 above)
Comment (sonalkr132) [reply lib/elastic_searcher.rb:62]: spaces can be weird. `[[:space:]]` gets most of them.

File: app/assets/javascripts/search.js (hunk 8 above)
Comment (sonalkr132) [reply app/assets/javascripts/search.js:23]: okay, didn't notice you moved space to var. You can leave it here but ideally, separate PR for unrelated changes is the best.

**Comment by sonalkr132** [comment]
Please post screenshots of changes to the view.

**Comment by vachhanihpavan** [comment]
> Please post screenshots of changes to the view.

:computer: [Screenshots](https://gist.github.com/vachhanihpavan/78fdbb9d02756398f1678c839e0de83b)

File: lib/elastic_searcher.rb [hunk 10]
```diff
@@ -59,6 +58,9 @@ def search_definition # rubocop:disable Metrics/MethodLength
         end
       end
 
+      # only return gems that are not yanked unless filter is used
+      post_filter { term yanked: false } unless query_str.gsub(/[[:space:]]/, "").include? "yanked:true"
```
Comment (sonalkr132) [inline lib/elastic_searcher.rb:62]: Found an issue with this. Using existing query with yanked filter removed means, gems where search term exists in summary and description are also shown. We want to show yanked gem only if query matches name only.

Comment (vachhanihpavan) [reply lib/elastic_searcher.rb:62]: Ah, correct. Missed it.

Comment (sonalkr132) [reply lib/elastic_searcher.rb:62]: for name too, we only want to show if there is exact match I think this will need separate query.

**Comment by sonalkr132** [comment]
Hi, I reverted this. Sorry I didn't think over this more carefully and ended up suggesting using ES :sweat:  . We don't need to make changes to `lib/elastic_searcher.rb` for this.
Adding `Rubygem.where(name: "query")` in controller should be enough. Changes to view are good :+1:
